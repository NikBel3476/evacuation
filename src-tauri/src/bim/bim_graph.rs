use super::bim_tools::{bim_t_rust, bim_transit_t_rust, bim_zone_t_rust};
use libc::size_t;

/// https://www.techiedelight.com/implement-graph-data-structure-c
/// Data structure to store a graph object
#[repr(C)]
pub struct bim_graph_t {
	/// An array of pointers to Node to represent an adjacency list
	pub head: *mut *mut bim_node_t,
	pub node_count: size_t,
}

pub struct bim_graph_t_rust {
	pub head: Vec<Box<bim_node_t_rust>>,
}

/// Data structure to store adjacency list nodes of the graph
#[repr(C)]
pub struct bim_node_t {
	pub dest: size_t,
	/// edge id
	pub eid: size_t,
	pub next: *mut bim_node_t,
}

#[derive(Default, Clone)]
pub struct bim_node_t_rust {
	pub dest: usize,
	pub eid: usize,
	pub next: Option<Box<bim_node_t_rust>>,
}

/// Data structure to store a graph edge
#[repr(C)]
pub struct bim_edge_t {
	pub src: size_t,
	pub dest: size_t,
	pub id: size_t,
}

pub struct bim_edge_t_rust {
	pub src: usize,
	pub dest: usize,
	pub id: usize,
}

pub fn bim_graph_new(bim: &bim_t_rust) -> bim_graph_t_rust {
	let edges = graph_create_edges(&bim.transits, &bim.zones);

	graph_create(&edges, edges.len(), bim.zones.len())
}

/// Function to create an adjacency list from specified edges
pub fn graph_create(
	edges: &[bim_edge_t_rust],
	edge_count: usize,
	node_count: usize,
) -> bim_graph_t_rust {
	// initialize head pointer for all vertices
	let mut graph_head: Vec<Box<bim_node_t_rust>> = vec![Default::default(); node_count];

	// add edges to the directed graph one by one
	for i in 0..edge_count {
		let edge = &edges[i];
		// get the source and destination vertex
		let src = edge.src;
		let dest = edge.dest;
		let eid = edge.id;

		// 1. allocate a new node of adjacency list from `src` to `dest`
		let mut new_node = bim_node_t_rust {
			dest,
			eid,
			next: None,
		};

		// point new node to the current head
		new_node.next = Some(graph_head[src].clone());

		// point head pointer to the new node
		graph_head[src] = Box::new(new_node);

		// 2. allocate a new node of adjacency list from `dest` to `src`
		let new_node_dest_to_src = bim_node_t_rust {
			dest: src,
			eid,
			// point new node to the current head
			next: Some(graph_head[dest].clone()),
		};

		// change head pointer to point to the new node
		graph_head[dest] = Box::new(new_node_dest_to_src);
	}

	// allocate storage for the graph data structure
	bim_graph_t_rust { head: graph_head }
}

pub fn graph_create_edges(
	list_doors: &[bim_transit_t_rust],
	zones: &[bim_zone_t_rust],
) -> Vec<bim_edge_t_rust> {
	let mut edges: Vec<bim_edge_t_rust> = vec![];

	for (i, transition) in list_doors.iter().enumerate() {
		let mut ids = [0, zones.len()];
		let mut j = 0usize;
		for (k, zone) in zones.iter().enumerate() {
			if equal_callback(zone, transition) && j != 2 {
				ids[j] = k;
				j += 1;
			}
		}

		let edge = bim_edge_t_rust {
			id: i,
			src: ids[0],
			dest: ids[1],
		};
		edges.push(edge);
	}

	edges
}

pub fn equal_callback(zone: &bim_zone_t_rust, transit: &bim_transit_t_rust) -> bool {
	for output in &zone.outputs {
		if output.eq(&transit.uuid) {
			return true;
		}
	}

	false
}
