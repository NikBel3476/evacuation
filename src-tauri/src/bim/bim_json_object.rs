use super::bim_polygon_tools;
use super::json_object::parse_building_from_json;
use libc::{c_char, c_double, c_ulonglong};

/// Количество символов в UUID + NUL символ
#[derive(Clone)]
#[repr(C)]
pub struct uuid_t {
	pub x: [c_char; 36 + 1],
}

impl Default for uuid_t {
	fn default() -> Self {
		Self { x: [0; 36 + 1] }
	}
}

#[repr(C)]
pub struct point_t_rust {
	pub x: c_double,
	pub y: c_double,
}

pub struct polygon_t_rust {
	pub numofpoints: c_ulonglong,
	pub points: *mut point_t_rust,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub enum BimElementSign {
	/// Указывает, что элемент здания является помещением/комнатой
	ROOM,
	/// Указывает, что элемент здания является лестницей
	STAIRCASE,
	/// Указывает, что элемент здания является проемом (без дверного полотна)
	DOOR_WAY,
	/// Указывает, что элемент здания является дверью, которая соединяет два элемента: ROOM и ROOM или ROOM и STAIR
	DOOR_WAY_IN,
	/// Указывает, что элемент здания является эвакуационным выходом
	DOOR_WAY_OUT,
	/// Указывает, что элемент является зоной вне здания
	OUTSIDE,
	/// Указывает, что тип элемента не определен
	#[default]
	UNDEFINED,
}

/// Структура, описывающая элемент
pub struct BimJsonElement {
	/// [JSON] UUID идентификатор элемента
	pub uuid: String,
	/// [JSON] Название элемента
	pub name: String,
	/// [JSON] Полигон элемента
	pub polygon: bim_polygon_tools::polygon_t_rust,
	/// [JSON] Массив UUID элементов, которые являются соседними к элементу
	pub outputs: Vec<String>,
	/// Внутренний номер элемента (генерируется)
	pub id: u64,
	/// [JSON] Количество людей в элементе
	pub number_of_people: u64,
	/// [JSON] Высота элемента
	pub size_z: f64,
	/// Уровень, на котором находится элемент
	pub z_level: f64,
	/// [JSON] Тип элемента
	pub sign: BimElementSign,
}

/// Структура поля, описывающего географическое положение объекта
pub struct BimJsonAddress {
	/// [JSON] Название города
	pub city: String,
	/// [JSON] Название улицы
	pub street_address: String,
	/// [JSON] Дополнительная информация о местоположении объекта
	pub add_info: String,
}

/// Структура, описывающая этаж
pub struct BimJsonLevel {
	/// [JSON] Название этажа
	pub name: String,
	/// [JSON] Массив элементов, которые принадлежат этажу
	pub build_elements: Vec<BimJsonElement>,
	/// [JSON] Высота этажа над нулевой отметкой
	pub z_level: f64,
}

/// Структура, описывающая здание
pub struct BimJsonObject {
	/// [JSON] Информация о местоположении объекта
	pub address: BimJsonAddress,
	/// [JSON] Название здания
	pub building_name: String,
	/// [JSON] Массив уровней здания
	pub levels: Vec<BimJsonLevel>,
}

pub fn bim_json_object_new(path_to_file: &str) -> BimJsonObject {
	let building = parse_building_from_json(path_to_file)
		.unwrap_or_else(|e| panic!("Failed to parse building. Error: {e}"));
	let mut bim_element_rs_id: u64 = 0;
	let mut bim_element_d_id: u64 = 0;

	let json_object = BimJsonObject {
		address: BimJsonAddress {
			city: building.address.city,
			street_address: building.address.street_address,
			add_info: building.address.add_info,
		},
		building_name: building.building_name,
		levels: building
			.levels
			.iter()
			.map(|level| BimJsonLevel {
				name: level.name.clone(),
				z_level: level.z_level,
				build_elements: level
					.build_elements
					.iter()
					.map(|element| BimJsonElement {
						uuid: element.id.clone(),
						name: element.name.clone(),
						id: match element.sign.as_str() {
							"Room" | "Staircase" => {
								let id = bim_element_rs_id;
								bim_element_rs_id += 1;
								id
							}
							"DoorWay" | "DoorWayInt" | "DoorWayOut" => {
								let id = bim_element_d_id;
								bim_element_d_id += 1;
								id
							}
							element_type => {
								panic!("Неизвестный тип элемента здания: {}", element_type)
							}
						},
						size_z: element.size_z,
						z_level: level.z_level,
						number_of_people: element.number_of_people,
						sign: match element.sign.as_str() {
							"Room" => BimElementSign::ROOM,
							"Staircase" => BimElementSign::STAIRCASE,
							"DoorWay" => BimElementSign::DOOR_WAY,
							"DoorWayInt" => BimElementSign::DOOR_WAY_IN,
							"DoorWayOut" => BimElementSign::DOOR_WAY_OUT,
							_ => BimElementSign::UNDEFINED,
						},
						outputs: element.outputs.clone(),
						polygon: bim_polygon_tools::polygon_t_rust {
							points: element.xy[0].points.clone(),
						},
					})
					.collect::<Vec<BimJsonElement>>(),
			})
			.collect::<Vec<BimJsonLevel>>(),
	};

	json_object
}
