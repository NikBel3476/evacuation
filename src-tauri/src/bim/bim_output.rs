use super::bim_tools::bim_t_rust;
use std::fs::File;
use std::io::Write;
use std::ops::Add;
use std::path::Path;

const OUTPUT_DIR: &str = "result";
pub const OUTPUT_DETAIL_FILE_RUST: &str = "_detailed_rust";
pub const OUTPUT_SHORT_FILE_RUST: &str = "_short_rust";
pub const OUTPUT_SUFFIX: &str = ".csv";

pub fn bim_basename_rust(path_to_file: &str) -> String {
	let basename = Path::new(Path::new(path_to_file).file_name().unwrap())
		.file_stem()
		.unwrap()
		.to_str()
		.unwrap()
		.to_owned();

	Path::new("..")
		.join(OUTPUT_DIR)
		.join(basename)
		.to_str()
		.unwrap()
		.to_owned()
}

pub fn bim_create_file_name_rust(base_file_name: &str, middle_name: &str, suffix: &str) -> String {
	Path::new(
		base_file_name
			.to_string()
			.add(middle_name)
			.add(suffix)
			.as_str(),
	)
	.to_str()
	.unwrap()
	.to_owned()
}

pub fn bim_output_head(bim: &bim_t_rust, file: &mut File) {
	file.write_all(b"t,").expect("Failed to write to file");

	for zone in &bim.zones {
		file.write_all(format!("{},", zone.name).as_bytes())
			.expect("Failed to write zone name to file");
	}

	for transition in &bim.transits {
		file.write_all(format!("{},", transition.name).as_bytes())
			.expect("Failed to write transition name to file");
	}

	file.write_all(b"\n").expect("Failed to write to file");
	file.flush().expect("Failed to flush file");
}

pub fn bim_output_body(bim: &bim_t_rust, time: f64, file: &mut File) {
	file.write_all(format!("{time:.2},").as_bytes())
		.expect("Failed to write to file");

	for zone in &bim.zones {
		file.write_all(format!("{:.2},", zone.number_of_people).as_bytes())
			.expect("Failed to write zone number of people to file");
	}

	for transition in &bim.transits {
		file.write_all(format!("{:.2},", transition.no_proceeding).as_bytes())
			.expect("Failed to write transition no_proceeding to file");
	}

	file.write_all(b"\n").expect("Failed to write to file");
	file.flush().expect("Failed to flush file");
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	#[cfg(any(target_os = "linux", target_os = "macos"))]
	fn test_bim_basename_linux_and_macos() {
		let path_ptr = "../res/two_levels.json";
		let out_file = bim_basename(path_ptr);
		let expected_path = "../result/two_levels";

		assert_eq!(expected_path, out_file);
	}

	#[test]
	#[cfg(target_os = "windows")]
	fn test_bim_basename_windows() {
		let path_ptr = "../res/two_levels.json";
		let out_file = bim_basename_rust(path_ptr);
		let expected_path = "..\\result\\two_levels";

		assert_eq!(expected_path, out_file);
	}
}
