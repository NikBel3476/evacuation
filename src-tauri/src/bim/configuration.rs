use serde::{Deserialize, Serialize};
use std::fs;
use std::path::Path;
use uuid::Uuid;

#[derive(Serialize, Deserialize, Eq, PartialEq)]
#[serde(rename_all(deserialize = "snake_case"))]
pub enum DistributionType {
	FromBim,
	Uniform,
}

#[derive(Serialize, Deserialize, Eq, PartialEq)]
#[serde(rename_all(deserialize = "snake_case"))]
pub enum TransitionType {
	FromBim,
	Users,
}

#[derive(Serialize, Deserialize)]
pub struct DistributionSpecial {
	pub uuid: Vec<Uuid>,
	pub density: f64,
	#[serde(rename(deserialize = "_comment"))]
	pub comment: String,
}

#[derive(Serialize, Deserialize)]
pub struct TransitionSpecial {
	pub uuid: Vec<Uuid>,
	pub width: f64,
	#[serde(rename(deserialize = "_comment"))]
	pub comment: String,
}

#[derive(Serialize, Deserialize)]
pub struct Distribution {
	#[serde(rename(deserialize = "type"))]
	pub distribution_type: DistributionType,
	pub density: f64,
	pub special: Vec<DistributionSpecial>,
}

#[derive(Serialize, Deserialize)]
pub struct Transition {
	#[serde(rename(deserialize = "type"))]
	pub transitions_type: TransitionType,
	#[serde(rename(deserialize = "doorwayin"))]
	pub doorway_in: f64,
	#[serde(rename(deserialize = "doorwayout"))]
	pub doorway_out: f64,
	pub special: Vec<TransitionSpecial>,
}

#[derive(Serialize, Deserialize)]
pub struct Modeling {
	pub step: f64,
	#[serde(rename(deserialize = "speed_max"))]
	pub max_speed: f64,
	#[serde(rename(deserialize = "density_max"))]
	pub max_density: f64,
	#[serde(rename(deserialize = "density_min"))]
	pub min_density: f64,
}

#[derive(Serialize, Deserialize)]
pub struct ScenarioCfg {
	#[serde(rename(deserialize = "bim"))]
	pub files: Vec<String>,
	#[serde(rename(deserialize = "logger_configure"))]
	pub logger_config: String,
	pub distribution: Distribution,
	#[serde(rename(deserialize = "transits"))]
	pub transition: Transition,
	pub modeling: Modeling,
}

pub fn load_cfg(path_to_file: &str) -> Result<ScenarioCfg, String> {
	match Path::new(path_to_file).exists() {
		true => {
			let json_content = fs::read_to_string(path_to_file).unwrap_or_else(|err| {
				panic!(
					"Ошибка чтения файла конфигурации сценария {}: {}",
					path_to_file, err
				)
			});

			let cfg: ScenarioCfg = serde_json::from_str(&json_content).unwrap_or_else(|err| {
				panic!(
					"Ошибка десериализации файла конфигурации сценария {}: {}",
					path_to_file, err
				)
			});

			Ok(cfg)
		}
		false => Err(format!("Не удалось найти указанный файл: {}", path_to_file)),
	}
}
