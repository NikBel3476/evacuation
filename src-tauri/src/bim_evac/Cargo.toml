[package]
name = "bim_evac"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
libc = "0.2"
bim_json_object = { path = "../../../src/bim_json_object" }
bim_tools = { path = "../../../src/bim_tools" }
bim_graph = { path = "../bim_graph" }
bim_polygon_tools = { path = "../../../src/bim_polygon_tools" }

[lib]
crate-type=["staticlib", "lib"]