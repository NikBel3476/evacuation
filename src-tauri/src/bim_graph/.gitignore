bim_graph/Cargo.lock
target