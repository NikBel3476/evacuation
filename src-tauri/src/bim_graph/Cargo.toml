[package]
name = "bim_graph"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
libc = "0.2"
bim_tools = { path = "../../../src/bim_tools" }

[lib]
crate-type=["staticlib", "lib"]