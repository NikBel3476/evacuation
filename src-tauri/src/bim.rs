use bim_evac::{
	evac_def_modeling_step, evac_moving_step_test_with_log, evac_moving_step_test_with_log_rust,
	evac_set_density_max_rust, evac_set_density_min_rust, evac_set_modeling_step_rust,
	evac_set_speed_max_rust, get_time_m, get_time_s, set_density_max, set_density_min,
	set_modeling_step, set_speed_max, time_inc, time_reset,
};
use bim_graph::{bim_graph_new, bim_graph_new_test};
use bim_json_object::{bim_json_object_new, BimElementSign};
use bim_output::{
	bim_basename_rust, bim_create_file_name_rust, bim_output_body, bim_output_head,
	OUTPUT_DETAIL_FILE_RUST, OUTPUT_SHORT_FILE_RUST, OUTPUT_SUFFIX,
};
use bim_tools::{bim_t_rust, bim_tools_get_num_of_people, bim_tools_new_rust, set_people_to_zone};
use cli::CliParameters;
use configuration::{load_cfg, DistributionType, ScenarioCfg, TransitionType};
use std::io::Write;

mod bim_cli;
mod bim_configure;
mod bim_evac;
mod bim_graph;
mod bim_json_object;
mod bim_output;
mod bim_polygon_tools;
mod bim_tools;
mod cli;
pub mod configuration;
mod graph;

pub fn run_rust() {
	// TODO: remove mock file path
	let cli_parameters = CliParameters {
		scenario_file: String::from("../scenario.json"),
	};

	let scenario_configuration = load_cfg(&cli_parameters.scenario_file)
		.unwrap_or_else(|e| panic!("Error reading the scenario configuration file. Error: {e}"));

	// TODO: add the logger

	for file in &scenario_configuration.files {
		let filename = bim_basename_rust(file);
		let log_filename = bim_basename_rust("log.txt");

		// Чтение файла и разворачивание его в структуру
		let bim_json = bim_json_object_new(file);

		let mut bim = bim_tools_new_rust(&bim_json);

		applying_scenario_bim_params(&mut bim, &scenario_configuration);

		// Files with results
		let output_detail =
			bim_create_file_name_rust(&filename, OUTPUT_DETAIL_FILE_RUST, OUTPUT_SUFFIX);
		let output_short =
			bim_create_file_name_rust(&filename, OUTPUT_SHORT_FILE_RUST, OUTPUT_SUFFIX);
		let log = bim_create_file_name_rust(&log_filename, "_rust", ".txt");

		let mut fp_detail =
			std::fs::File::create(&output_detail).expect("Error opening the output file");
		let mut fp_short =
			std::fs::File::create(&output_short).expect("Error opening the output file");
		let mut log_file = match std::path::Path::new(&log).exists() {
			true => std::fs::File::options()
				.append(true)
				.open(&log)
				.unwrap_or_else(|e| panic!("Error opening the log file. Error: {e}")),
			false => std::fs::File::create(&log)
				.unwrap_or_else(|e| panic!("Error create the log file. Error: {e}")),
		};

		let current_time = chrono::Local::now()
			.format("%Y-%m-%d %H:%M:%S.%6f")
			.to_string();
		let filename_log = format!("The file name of the used bim `{filename}.json`\n");
		print!("{current_time} {filename_log}");
		log_file
			.write_all(filename_log.as_bytes())
			.unwrap_or_else(|e| panic!("Failed to write log to file. Error: {e}"));

		bim_output_head(&bim, &mut fp_detail);
		bim_output_body(&bim, 0.0, &mut fp_detail);

		// let graph = bim_graph_new_rust(&bim);
		let graph = bim_graph_new(&bim);
		// let bim_graph = bim_graph_new_test(&bim);
		// TODO: add print graph

		evac_def_modeling_step(&bim);
		time_reset();

		let remainder = 0.0; // Количество человек, которое может остаться в зд. для остановки цикла
		loop {
			// evac_moving_step_test_with_log(bim_graph, &mut bim.zones, &mut bim.transits);
			evac_moving_step_test_with_log_rust(&graph, &mut bim.zones, &mut bim.transits);
			time_inc();
			bim_output_body(&bim, get_time_m(), &mut fp_detail);

			let mut num_of_people = 0.0;
			for zone in &bim.zones {
				if zone.is_visited {
					num_of_people += zone.number_of_people;
				}
			}

			if num_of_people <= remainder {
				break;
			}
		}

		let num_of_evacuated_people = bim_tools_get_num_of_people(&bim);
		let evacuation_time_m = get_time_m();
		let evacuated_people = bim.zones[bim.zones.len() - 1].number_of_people;

		let evac_time_log = format!(
			"{current_time} Длительность эвакуации: {:.2} с. ({:.2} мин.)\n",
			get_time_s(),
			evacuation_time_m
		);
		let number_of_people_log = format!("{current_time} Количество человек: в здании - {num_of_evacuated_people:.2} (в безопасной зоне - {evacuated_people:.2}) чел.\n");
		let delimiter = format!("{current_time} ---------------------------------------\n");

		print!("{evac_time_log}");
		log_file
			.write_all(evac_time_log.as_bytes())
			.unwrap_or_else(|e| panic!("Failed to write log to file. Error: {e}"));
		print!("{number_of_people_log}");
		log_file
			.write_all(number_of_people_log.as_bytes())
			.unwrap_or_else(|e| panic!("Failed to write log to file. Error: {e}"));
		print!("{delimiter}");
		log_file
			.write_all(delimiter.as_bytes())
			.unwrap_or_else(|e| panic!("Failed to write log to file. Error: {e}"));
		log_file
			.flush()
			.unwrap_or_else(|e| panic!("Failed to flush log to file. Error: {e}"));

		fp_short
			.write_all(
				format!(
					"{evacuation_time_m:.2},{num_of_evacuated_people:.2},{evacuated_people:.2}\n"
				)
				.as_bytes(),
			)
			.unwrap_or_else(|e| panic!("Failed to write fp_short to file. Error: {e}"));
		fp_short
			.flush()
			.unwrap_or_else(|e| panic!("Failed to flush fp_short to file. Error: {e}"));
	}
}

pub fn applying_scenario_bim_params(bim: &mut bim_t_rust, scenario_configuration: &ScenarioCfg) {
	for transition in &mut bim.transits {
		if scenario_configuration.transition.transitions_type == TransitionType::Users {
			match transition.sign {
				BimElementSign::DOOR_WAY_IN => {
					transition.width = scenario_configuration.transition.doorway_in
				}
				BimElementSign::DOOR_WAY_OUT => {
					transition.width = scenario_configuration.transition.doorway_out
				}
				_ => {}
			}
		}

		// A special set up the transit width of item of bim
		for special in &scenario_configuration.transition.special {
			for uuid in &special.uuid {
				if transition.uuid.eq(uuid) {
					transition.width = special.width;
				}
			}
		}
	}

	// in c code bim->transits is a pointers to bim->levels[_]->transits so necessary to update bim->levels[_]->transits
	for level in &mut bim.levels {
		for transition in &mut level.transits {
			if scenario_configuration.transition.transitions_type == TransitionType::Users {
				match transition.sign {
					BimElementSign::DOOR_WAY_IN => {
						transition.width = scenario_configuration.transition.doorway_in
					}
					BimElementSign::DOOR_WAY_OUT => {
						transition.width = scenario_configuration.transition.doorway_out
					}
					_ => {}
				}
			}

			// A special set up the transit width of item of bim
			for special in &scenario_configuration.transition.special {
				for uuid in &special.uuid {
					if transition.uuid.eq(uuid) {
						transition.width = special.width;
					}
				}
			}
		}
	}

	for zone in &mut bim.zones {
		if zone.sign == BimElementSign::OUTSIDE {
			continue;
		}

		if scenario_configuration.distribution.distribution_type == DistributionType::Uniform {
			set_people_to_zone(
				zone,
				zone.area * scenario_configuration.distribution.density,
			);
		}

		// A special set up the density of item of bim
		for special in &scenario_configuration.distribution.special {
			for uuid in &special.uuid {
				if zone.uuid.eq(uuid) {
					set_people_to_zone(zone, zone.area * special.density);
				}
			}
		}
	}

	// in c code bim->zones is a pointers to bim->levels[_]->zones so necessary to update bim->levels[_]->zones
	for level in &mut bim.levels {
		for zone in &mut level.zones {
			if zone.sign == BimElementSign::OUTSIDE {
				continue;
			}

			if scenario_configuration.distribution.distribution_type == DistributionType::Uniform {
				set_people_to_zone(
					zone,
					zone.area * scenario_configuration.distribution.density,
				);
			}

			// A special set up the density of item of bim
			for special in &scenario_configuration.distribution.special {
				for uuid in &special.uuid {
					if zone.uuid.eq(uuid) {
						set_people_to_zone(zone, zone.area * special.density);
					}
				}
			}
		}
	}

	set_modeling_step(scenario_configuration.modeling.step);
	evac_set_modeling_step_rust(scenario_configuration.modeling.step);
	set_speed_max(scenario_configuration.modeling.max_speed);
	evac_set_speed_max_rust(scenario_configuration.modeling.max_speed);
	set_density_max(scenario_configuration.modeling.max_density);
	evac_set_density_max_rust(scenario_configuration.modeling.max_density);
	set_density_min(scenario_configuration.modeling.min_density);
	evac_set_density_min_rust(scenario_configuration.modeling.min_density);
}
