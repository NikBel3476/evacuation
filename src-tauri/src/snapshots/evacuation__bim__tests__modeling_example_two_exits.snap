---
source: src/bim.rs
expression: modeling_result
---
number_of_people_in_building: 0
evacuation_time_in_seconds: 25.200000000000014
number_of_evacuated_people: 83.7347693024665

